libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp

```diff
@@ -0,0 +1,792 @@
+#include <SRV_Channel/SRV_Channel.h>
+#include <GCS_MAVLink/GCS.h>
+
+#include "AP_FETtecOneWire.h"
+#if HAL_AP_FETTEC_ONEWIRE_ENABLED
+
+extern const AP_HAL::HAL& hal;
+
+static constexpr uint32_t FTOW_UART_LOCK_KEY = 0x0FE77EC0;
+static constexpr uint32_t DELAY_TIME_US = 700;
+static constexpr uint32_t BAUDRATE = 500000;
+static constexpr uint8_t ALL_ID = 0x1F;
+static constexpr uint8_t FRAME_OVERHEAD = 6;
+static constexpr uint8_t MAX_TRANSMIT_LENGTH = 4;
+static constexpr uint8_t MAX_RECEIVE_LENGTH = 12;
+static constexpr uint8_t MAX_RESPONSE_LENGTH = FRAME_OVERHEAD + MAX_RECEIVE_LENGTH;
+
+const AP_Param::GroupInfo AP_FETtecOneWire::var_info[] = {
+    // @Param: MASK
+    // @DisplayName: Channel Bitmask
+    // @Description: Enable of FETtec OneWire ESC protocol to specific channels
+    // @Bitmask: 0:Channel1,1:Channel2,2:Channel3,3:Channel4,4:Channel5,5:Channel6,6:Channel7,7:Channel8,8:Channel9,9:Channel10,10:Channel11,11:Channel12,12:Channel13,13:Channel14,14:Channel15,15:Channel16
+    // @RebootRequired: True
+    // @User: Standard
+    AP_GROUPINFO("MASK",  1, AP_FETtecOneWire, _motor_mask, 0),
+
+    // @Param: POLES
+    // @DisplayName: Nr. electrical poles
+    // @Description: Number of motor electrical poles
+    // @Range: 2 50
+    // @RebootRequired: False
+    // @User: Standard
+    AP_GROUPINFO("POLES", 2, AP_FETtecOneWire, _pole_count, 14),
+
+    AP_GROUPEND
+};
+
+AP_FETtecOneWire *AP_FETtecOneWire::_singleton;
+
+AP_FETtecOneWire::AP_FETtecOneWire()
+{
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+    if (_singleton != nullptr) {
+        AP_HAL::panic("AP_FETtecOneWire must be singleton");
+    }
+#endif
+    _singleton = this;
+
+    _response_length[OW_OK] = 1;
+    _response_length[OW_BL_START_FW] = 0;       // BL only
+    _response_length[OW_REQ_TYPE] = 1;
+    _response_length[OW_REQ_SN] = 12;
+    _response_length[OW_REQ_SW_VER] = 2;
+    _response_length[OW_SET_FAST_COM_LENGTH] = 1;
+    _response_length[OW_SET_TLM_TYPE] = 1;
+}
+
+/**
+  initialize the serial port, scan the OneWire bus, setup the found ESCs
+*/
+void AP_FETtecOneWire::init()
+{
+    if (!_uart_initialised) {
+        AP_SerialManager& serial_manager = AP::serialmanager();
+        _uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FETtecOneWire, 0);
+        if (_uart) {
+            _uart->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
+            _uart->set_unbuffered_writes(true);
+            _uart->set_blocking_writes(false);
+            _uart->begin(BAUDRATE);
+            _uart_initialised = true;
+        }
+    }
+
+    if (_uart == nullptr) {
+        return;
+    }
+
+    const uint32_t now = AP_HAL::micros();
+
+    if (_scan_active < MOTOR_COUNT_MAX) {
+        if (now - _last_send_us < DELAY_TIME_US) {
+            // scan should not be done too fast, as the bootloader has some message timing requirements.
+            return;
+        } else {
+            _last_send_us = now;
+        }

```

Comment:
@teezeefpv is this really necessary? the loop period is 2500us and DELAY_TIME_US is 700us.